### PR TITLE
Verify, Confirm, Unverify Commands

### PR DESCRIPTION
**Pull request by zaydtahir**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #15 #16 #23 #29 

## What is the new behavior?

- Added a colorUtil file to hold common colors
- Added verify command
- Added unverify command
- Added confirm command
- Added adminverify command

## Other information

I know there's a lot of issues associated with but most of the commands all deal with verification so it made sense to do these in one PR

### CODE REVIEW

File: src/database/models/SettingsModel.ts [hunk 1]
```diff
@@ -1,15 +1,31 @@
 import {model, Schema} from 'mongoose';
 
+interface ISettings {
```
Comment (AadityaChaudhary) [inline src/database/models/SettingsModel.ts:3]: whats the I for? what is our naming convention?

**Review by AadityaChaudhary — APPROVED** [review]
assuming it works, LGTM

File: src/database/models/SettingsModel.ts (hunk 1 above)
Comment (zaydtahir) [reply src/database/models/SettingsModel.ts:3]: I stands for interface

**Review by jlsajfj — APPROVED** [review]
lgtm, note for the future: move email to external file

**Comment by zaydtahir** [comment]
> lgtm, note for the future: move email to external file

Created an issue for this

File: src/database/models/QuoteModel.ts [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
+  guildId: string,
+  quote: string,
+  author: string,
+}
+
 export const Quote = new Schema({
   guildId: String,
   quote: String,
   author: String,
 });
```
Comment (sahil-kale) [inline src/database/models/QuoteModel.ts:14]: Something i thought of just now, could we store date/time info?

File: src/database/models/SettingsModel.ts [hunk 3]
```diff
@@ -1,15 +1,31 @@
...
+  prefix: string,
+  timezone: string,
+  verificationEnabled: boolean,
+  pinsEnabled: boolean,
+  quotesEnabled: boolean,
+  birthdayAnnouncementsEnabled: boolean,
+  morningAnnouncementsEnabled: boolean,
+  dueDatesEnabled: boolean,
+  emailDomain: string,
+  verifiedRole: string,
```
Comment (sahil-kale) [inline src/database/models/SettingsModel.ts:14]: Forgetting admin role

File: src/tests/database/VerifiedUserModel.test.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    expect(await emailAddressLinkedToUser(emailAddress, guildId)).toBe(false);
+
+    await verifiedUserModel.create({
+      emailHash: await hashString(emailAddress),
+      guildId: guildId,
+    });
+
+    expect(await emailAddressLinkedToUser(emailAddress, guildId)).toBe(true);
+  });
+});
```
Comment (sahil-kale) [inline src/tests/database/VerifiedUserModel.test.ts:35]: Can you test that this test will fail if the guild ID is wrong?

File: src/tests/lib/BediEmbed.test.ts [hunk 5]
```diff
@@ -1,19 +1,13 @@
...
   });
-
-  test('BediEmbed has Correct Timestamp', () => {
-    const {timestamp} = new BediEmbed();
-
-    const now = Date.now();
-
-    expect(timestamp).toBe(now);
-  });
 });
```
Comment (sahil-kale) [inline src/tests/lib/BediEmbed.test.ts:13]: This is one of the few times I think something doesn't need to be tested, I would honestly just assert/expect that the embed != None

File: src/tests/utils/emailUtil.test.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+describe('Email Utils', () => {
+  test('isEmailValid', () => {
+    expect(isEmailValid('test@gmail.com')).toBe(true);
+    expect(isEmailValid('test@yahoo.ca')).toBe(true);
+    expect(isEmailValid('test@gmail')).toBe(false);
+    expect(isEmailValid('@gmail.com')).toBe(false);
+    expect(isEmailValid('testgmail.com')).toBe(false);
+    expect(isEmailValid('test@gmailcom')).toBe(false);
+    expect(isEmailValid('test@gmail.c')).toBe(false);
```
Comment (sahil-kale) [inline src/tests/utils/emailUtil.test.ts:11]: How is this not considered to be valid?

File: src/utils/discordUtil.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import {Message, Role} from 'discord.js';
+import logger from './loggerUtil';
+import {getSettings} from '../database/models/SettingsModel';
+
+export const addRoleToAuthor = async (message: Message, roleName: string) => {
```
Comment (sahil-kale) [inline src/utils/discordUtil.ts:5]: Hmmm, any chance we can make this generic and pass in a user ID we want to add a role to?

Comment (sahil-kale) [reply src/utils/discordUtil.ts:5]: I think it's kind of limiting on a function to only be able to add a role to the author when we can pass that in as an argument and then use the same function for verification etc

File: src/utils/emailUtil.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    text: text,
+    html: htmlText,
+    auth: {
+      accessToken: oAuth2Client.getAccessToken(),
+    },
+  };
+
+  transporter.sendMail(mailOptions, function(error: Error | null, info: SentMessageInfo) {
+    if (error) {
+      logger.error(error);
```
Comment (sahil-kale) [inline src/utils/emailUtil.ts:41]: Can we get an embed here that just tells the user there was a fatal error in processing this command? Something like "Yo, having trouble, sorry!". Or this could be the caller's responsibility, but your choice

**Review by sahil-kale — CHANGES_REQUESTED** [review]
Overall looks good, but there's a few things I'd like to see addressed
1) Brief descriptions t the top of each file detailing its general purpose. This PR was very hard to follow due to the large amount of stuff
2) I'm realzing this now, but we should have more detailed error handling than just a logger.error(), something to alert the user that the bot was not operational rather than it just going silent like old bedibot would be helpful

File: src/commands/General/ping.ts [hunk 9]
```diff
@@ -22,7 +22,6 @@ module.exports = class PingCommand extends Command {
     });
 
     const editEmbed = new BediEmbed()
-        .setColor('BLUE')
```
[the remark this answers is not in the document]
Comment (zaydtahir) [reply src/commands/General/ping.ts:25]: The BediEmbed object assigns the "PRIMARY" color to itself by default (found in colorUtil file)

File: src/commands/Verification/adminVerify.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+          .setColor(colors.ERROR)
+          .setTitle('Admin Verify Reply')
+          .setDescription('Verification is not enabled on this server!'); // TODO: Explain how to enable verification when implemented
+      return message.channel.send({
+        embeds: [embed],
+      });
+    }
+
+    const user = await args.pickResult('user');
+    if (!user.success) {
```
[the remark this answers is not in the document]
Comment (zaydtahir) [reply src/commands/Verification/adminVerify.ts:34]: This function override from Command. The IDE should be able to pick up the docstring as long as you have the dependencies installed.

[the remark this answers is not in the document]
Comment (sahil-kale) [reply src/commands/Verification/adminVerify.ts:34]: Ok, sounds good

File: src/database/models/SettingsModel.ts (hunk 3 above)
Comment (zaydtahir) [reply src/database/models/SettingsModel.ts:14]: Chose to check user permissions for that functionality instead.

File: src/commands/Verification/verify.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      description: 'Allows you to verify yourself and access the server',
+      preconditions: ['GuildOnly'],
+    });
+  }
+
+  async run(message: Message, args: Args) {
+    const {guild, guildId, author} = message;
+    const settingsData = await getSettings(guildId as string);
+
+    await message.delete();
```
[the remark this answers is not in the document]
Comment (zaydtahir) [reply src/commands/Verification/verify.ts:28]: Added a comment in the code to clarify this.

File: src/commands/Verification/adminVerify.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+  async run(message: Message, args: Args) {
+    const {guild, guildId, author} = message;
+    const settingsData = await getSettings(guildId as string);
+
+    if (!settingsData.verificationEnabled) {
+      const embed = new BediEmbed()
+          .setColor(colors.ERROR)
+          .setTitle('Admin Verify Reply')
+          .setDescription('Verification is not enabled on this server!'); // TODO: Explain how to enable verification when implemented
```
Comment (sahil-kale) [inline src/commands/Verification/adminVerify.ts:28]: Can this be made an issue please?

File: src/listeners/commandDenied.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    const embed = new BediEmbed()
+        .setTitle(capFirstLetter(command.name) + ' Reply')
+        .setColor(colors.ERROR)
+        .setDescription(content);
+
+    return message.reply({
+      embeds: [embed],
+    });
+  }
+};
```
Comment (sahil-kale) [inline src/listeners/commandDenied.ts:26]: Can we also add something for the logger here as well?

**Review by sahil-kale — APPROVED** [review]
